### PR TITLE
Java bindings for nans to nulls [skip ci]

### PR DESCRIPTION
**Pull request by razajafri**
This PR adds Java bindings for converting nans to nulls

### CODE REVIEW

File: java/src/main/java/ai/rapids/cudf/ColumnVector.java [hunk 1]
```diff
@@ -207,6 +207,14 @@ private synchronized ColumnVector incRefCountInternal(boolean isFirstTime) {
     return this;
   }
 
+  /**
+   * Returns a new ColumnVector based on 'this' with NaNs converted to nulls, preserving the
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ColumnVector.java:211]: Nit: `based on 'this'` is redundant given this is an instance method.

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 2]
```diff
@@ -1291,4 +1291,19 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_title(JNIEnv *env, jobj
   CATCH_STD(env, 0);
 }
 
+JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_nansToNulls(JNIEnv *env, jobject j_object, jlong handle) {
+
+  JNI_NULL_CHECK(env, handle, "native view handle is null", 0)
+
+  try {
```
Comment (jlowe) [inline java/src/main/native/src/ColumnVectorJni.cpp:1298]: All JNI entry-point methods must call `auto_set_device` to set the CUDA device, if necessary.
```suggestion
  try {
    cudf::jni::auto_set_device(env);
```

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 3]
```diff
@@ -1291,4 +1291,19 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_title(JNIEnv *env, jobj
...
+    cudf::column_view * view = reinterpret_cast<cudf::column_view *>(handle);
+    auto pair = cudf::experimental::nans_to_nulls(*view);
+    std::unique_ptr<cudf::column> result(new cudf::column(*view));
```
Comment (jlowe) [inline java/src/main/native/src/ColumnVectorJni.cpp:1301]: This makes a copy of the original column's null mask which we then turn around and smash with the `set_null_mask` call below.  We should avoid making this wasteful allocation and copy.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 4]
```diff
@@ -2003,4 +2003,21 @@ void testStringTitlize() {
       assertColumnsAreEqual(expected, result);
     }
   }
+
+  @Test
+  void testNansToNulls() {
+    try (ColumnVector cvFloat = ColumnVector.fromBoxedFloats(1.2f, Float.POSITIVE_INFINITY,
```
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:2009]: I recommend moving the list of floats outside of this try-with-resources block to make it much easier to read, e.g.:
```java
  Float[] inputFloats = new Float[] { 1.2f, Float.POSITIVE_INFINITY, ... };
  Float[] expectedFloats = new Float[] { 1.2f, ... };
  ...
  try(ColumnVector cvFloat = ColumnVector.fromBoxedFloats(inputFloats);
      ColumnVector expectedFloat = ColumnVector.fromBoxedFloats(expectedFloats);
        ...) {
```

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 5]
```diff
@@ -2003,4 +2003,21 @@ void testStringTitlize() {
...
+        Float.NEGATIVE_INFINITY, null, Float.NaN, Float.MAX_VALUE, Float.MIN_VALUE, 435243.2323f);
+         ColumnVector cvDouble = ColumnVector.fromBoxedDoubles(1.2d, Double.POSITIVE_INFINITY,
```
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:2011]: There should be values of other bit representation for NaN (-NaN and at least one other IEEE 754 NaN value that isn't the canonical NaN value produced by Float.NaN).

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 6]
```diff
@@ -1291,4 +1291,20 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_title(JNIEnv *env, jobj
...
 
+JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_nansToNulls(JNIEnv *env, jobject j_object, jlong handle) {
+
+  JNI_NULL_CHECK(env, handle, "native view handle is null", 0)
+
+  try {
+    cudf::jni::auto_set_device(env);
+    cudf::column_view * view = reinterpret_cast<cudf::column_view *>(handle);
+    auto pair = cudf::experimental::nans_to_nulls(*view);
+    cudf::bitmask_type const* data = reinterpret_cast<cudf::bitmask_type const*>(pair.first->data());
```
Comment (revans2) [inline java/src/main/native/src/ColumnVectorJni.cpp:1302]: You did what @jlowe asked, but the code is kind of hard to follow.  A comment explaining what you are doing and why would be helpful.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 7]
```diff
@@ -2003,4 +2003,36 @@ void testStringTitlize() {
       assertColumnsAreEqual(expected, result);
     }
   }
+
+  @Test
+  void testNansToNulls() {
+    Float[] floats = new Float[]{1.2f, Float.POSITIVE_INFINITY,
+        Float.NEGATIVE_INFINITY, null, -Float.NaN, Float.NaN, Float.MAX_VALUE, Float.MIN_VALUE,
```
Comment (revans2) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:2010]: `-Float.NaN` does not produce a negative NaN. Java will apply the negation operation to the NaN and end up with just another NaN.  You have to jump through some special hoops to get one. i.e. Create the bit representation of it in an int and call `Float.intBitsToFloat`.  There is a Double version with long as well.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 8]
```diff
@@ -2003,4 +2003,36 @@ void testStringTitlize() {
...
+        435243.2323f, Float.intBitsToFloat(0x7f800001), Float.intBitsToFloat(0x7fffffff),
+        Float.intBitsToFloat(0xff800001), Float.intBitsToFloat(0xffffffff)};
+
+    Double[] doubles = new Double[]{1.2d, Double.POSITIVE_INFINITY,
+        Double.NEGATIVE_INFINITY, null, Double.NaN, -Double.NaN, Double.MAX_VALUE,
```
Comment (revans2) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:2015]: Here too -NaN does not do what you want.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 9]
```diff
@@ -2003,4 +2003,36 @@ void testStringTitlize() {
...
+        Double.MIN_VALUE, 435243.2323d,
+        Double.longBitsToDouble(0x7ff0000000000001L), Double.longBitsToDouble(0x7fffffffffffffffL),
```
Comment (revans2) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:2017]: If these are supposed to be special double values, let's make them constants somewhere so everyone knows what they are and can reuse them without some special magic.

File: java/src/main/native/src/ColumnVectorJni.cpp (hunk 6 above)
Comment (jlowe) [reply java/src/main/native/src/ColumnVectorJni.cpp:1302]: I also recommend calling this local "mask" or "validity" or something to indicate what it is.  Just calling it `data` makes it easy to confuse with the data vector.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 10]
```diff
@@ -2003,4 +2003,36 @@ void testStringTitlize() {
...
+        Double.longBitsToDouble(0xfff0000000000001L), Double.longBitsToDouble(0xffffffffffffffffL)};
+
+    Float[] expectedFloats = new Float[]{1.2f, Float.POSITIVE_INFINITY,
```
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:2020]: Nit: It's much easier to cross-check the input values with the expected values when they're right next to each other, so I recommend grouping these by type rather than all input and then all expected.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java (hunk 7 above)
Comment (razajafri) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: Thanks for pointing out the -Float.NaN. It is borderline embarrassing! 

I have used the intBitsToFloat in this same test. Not sure if you didn't look at it or are referring to something different from what I have used.

File: java/src/main/native/src/ColumnVectorJni.cpp [hunk 11]
```diff
@@ -1291,4 +1291,24 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_title(JNIEnv *env, jobj
...
+    cudf::jni::auto_set_device(env);
+    cudf::column_view * view = reinterpret_cast<cudf::column_view *>(handle);
+    // get a new null mask by setting all the nans to null
+    std::pair<std::unique_ptr<rmm::device_buffer>, cudf::size_type> pair = cudf::experimental::nans_to_nulls(*view);
+    // create a column_view which is a no-copy wrapper around the original column without the null mask
+    std::unique_ptr<cudf::column_view> copy_view(new cudf::column_view(view->type(), view->size(), view->data<char>()));
+    // create a column by deep copying the copy_view
+    std::unique_ptr<cudf::column> copy(new cudf::column(*copy_view));
+    // set the null mask with nans set to null
+    copy->set_null_mask(std::move(*pair.first), pair.second);
```
Comment (jlowe) [inline java/src/main/native/src/ColumnVectorJni.cpp:1308]: Don't we need to call `release` on the `unique_ptr` here to transfer ownership of the `device_buffer`?

Comment (razajafri) [reply java/src/main/native/src/ColumnVectorJni.cpp:1308]: It's not needed as per the docs https://en.cppreference.com/w/cpp/memory/unique_ptr/operator%3D

Comment (jlowe) [reply java/src/main/native/src/ColumnVectorJni.cpp:1308]: But we're not using `unique_ptr`'s assignment operator here.  We're invoking `std::move` on the result of `unique_ptr`'s dereference operator, which doesn't inform `unique_ptr` of what's going on.  It will still think it owns the `device_buffer` and destroy it when the `pair` variable falls out of scope at the end of this method.

It think this needs to be something like:
```c++
  copy->set_null_mask(std::move(*(pair.first.release())), pair.second);
```

Comment (razajafri) [reply java/src/main/native/src/ColumnVectorJni.cpp:1308]: aah. I see. Thanks for pointing this out.

**Comment by razajafri** [comment]
@jlowe @revans2 
Does this look OK to go in?

**Comment by jlowe** [comment]
No, need to resolve [this discussion](https://github.com/rapidsai/cudf/pull/4992#discussion_r414736120) by reverting 1d9857bf7a5e9c8319f88f0acbe2c63db8cc66e6

**Comment by razajafri** [comment]
Reverted

**Comment by razajafri** [comment]
@kkraus14 this is ready to go in
